ibm/resource_ibm_pi_dhcp.go

```diff
@@ -0,0 +1,236 @@
+	sess, err := meta.(ClientSession).IBMPISession()
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	cloudInstanceID := d.Get(helpers.PICloudInstanceId).(string)
+
+	client := st.NewIBMPIDhcpClient(sess, cloudInstanceID)
+	dhcpServer, err := client.CreateWithContext(ctx, cloudInstanceID)
+	if err != nil {
+		log.Printf("[DEBUG] create DHCP failed %v", err)
+		return diag.Errorf(errors.CreateDchpOperationFailed, cloudInstanceID, err)
+	}
+
+	d.SetId(fmt.Sprintf("%s/%s", cloudInstanceID, *dhcpServer.ID))
+
+	_, err = waitForIBMPIDhcpStatus(ctx, client, *dhcpServer.ID, cloudInstanceID, d.Timeout(schema.TimeoutCreate))
+	if err != nil {
+		return diag.Errorf(errors.CreateDchpOperationFailed, cloudInstanceID, err)
+	}
+
+	return resourceIBMPIDhcpRead(ctx, d, meta)
+}
+
+func resourceIBMPIDhcpRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	sess, err := meta.(ClientSession).IBMPISession()
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	parts, err := idParts(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	cloudInstanceID := parts[0]
+	dhcpID := parts[1]
+
+	client := st.NewIBMPIDhcpClient(sess, cloudInstanceID)
+	dhcpServer, err := client.GetWithContext(ctx, dhcpID, cloudInstanceID)
+	if err != nil {
+		switch err.(type) {
+		case *p_cloud_service_d_h_c_p.PcloudDhcpGetNotFound:
+			log.Printf("[DEBUG] dhcp does not exist %v", err)
+			d.SetId("")
+			return nil
+		}
+		log.Printf("[DEBUG] get DHCP failed %v", err)
+		return diag.Errorf(errors.GetDhcpOperationFailed, dhcpID, err)
+	}
+
+	d.Set(PIDhcpId, *dhcpServer.ID)
+	d.Set(PIDhcpStatus, *dhcpServer.Status)
+
+	if dhcpServer.Network != nil {
+		d.Set(PIDhcpNetwork, dhcpServer.Network.ID)
+	}
+	if dhcpServer.Leases != nil {
+		leaseList := make([]map[string]string, len(dhcpServer.Leases))
+		for i, lease := range dhcpServer.Leases {
+			leaseList[i] = map[string]string{
+				PIDhcpInstanceIp:  *lease.InstanceIP,
+				PIDhcpInstanceMac: *lease.InstanceMacAddress,
+			}
+		}
+		d.Set(PIDhcpLeases, leaseList)
+	}
+
+	return nil
+}
+func resourceIBMPIDhcpDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	sess, err := meta.(ClientSession).IBMPISession()
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	parts, err := idParts(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	cloudInstanceID := parts[0]
+	dhcpID := parts[1]
+
+	client := st.NewIBMPIDhcpClient(sess, cloudInstanceID)
+	_, err = client.DeleteWithContext(ctx, dhcpID, cloudInstanceID)
+	if err != nil {
+		switch err.(type) {
+		case *p_cloud_service_d_h_c_p.PcloudDhcpDeleteNotFound:
+			log.Printf("[DEBUG] dhcp does not exist %v", err)
+			d.SetId("")
+			return nil
+		}
+		log.Printf("[DEBUG] delete DHCP failed %v", err)
+		return diag.Errorf(errors.DeleteDhcpOperationFailed, dhcpID, err)
+	}
+	waitForIBMPIDhcpDeleted(ctx, client, dhcpID, cloudInstanceID, d.Timeout(schema.TimeoutDelete))
```

Comment:
waitForIBMPIDhcpDeleted returns error as one argument can we read the error and retrun the error if its not nil